### PR TITLE
misc_services: Don't gossip hit rates when they are not used

### LINKED ISSUE #5183: Don't gossip cache hit rates when they are not used
When cache_hit_rate_read_balancing is set to false, heat-weighted load balancing is turned off and we gossip cache hit rates for no good reason. The size of gossiped data is linear on number of tables in the database so it's good to avoid such traffic when it is not used.

Ref https://github.com/scylladb/scylla-enterprise/issues/1092

### PR DESCRIPTION
**Pull request by haaawk**
When cache_hit_rate_read_balancing is set to false,
heat-weighted load balancing is turned off and we
gossip cache hit rates for no good reason.
The size of gossiped data is linear on number of tables
in the database so it's good to avoid such traffic
when it is not used.

Fixes #5183

Signed-off-by: Piotr Jastrzebski piotr@scylladb.com

### CODE REVIEW

**Comment by haaawk** [comment]
I've tested it the same way as before with 3 nodes cluster and a rolling disable and then enable of HWLB. Nothing bad happened.

File: main.cc [hunk 1]
```diff
@@ -1027,12 +1027,18 @@ int main(int ac, char** av) {
                 lb->stop_broadcasting().get();
             });
             supervisor::notify("starting cf cache hit rate calculator");
-            cf_cache_hitrate_calculator.start(std::ref(db), std::ref(cf_cache_hitrate_calculator)).get();
-            auto stop_cache_hitrate_calculator = defer_with_log_on_error([&cf_cache_hitrate_calculator] {
+
+            auto stop_cache_hitrate_calculator = defer_with_log_on_error([&cf_cache_hitrate_calculator, &db] {
```
Comment (tgrabiec) [inline main.cc]: This should be constructed after `cf_cache_hitrate_calculator.start()`, becasue if start() fails, we must not attempt to stop.

Comment (haaawk) [reply main.cc]: Ok. Sending fix now.

**Comment by haaawk** [comment]
@tgrabiec I moved the cleaner after the .start() is called.

**Comment by gleb-cloudius** [comment]
I am missing where do we check that entire cluster has the feature disabled?

**Comment by tgrabiec** [comment]
@gleb-cloudius It's disabled on per-request basis, if any of the replicas has the feature disabled. Changed in "filter_for_query: Don't use heat-weighted load balancing when".

**Comment by gleb-cloudius** [comment]
Ah, we skipping the calculation if one of the nodes returned DISABLE, right. Not sure this is enough though. With a cluster of size 4 if one node has it disable but three others are not some requests will be balanced and some will not which may result in disproportional load on some nodes.

**Comment by tgrabiec** [comment]
@gleb-cloudius So are you suggesting to disable the load balancing if any of the known and alive nodes have it disabled?

**Comment by gleb-cloudius** [comment]
On Wed, Oct 23, 2019 at 10:12:07AM -0700, Tomasz Grabiec wrote:
> @gleb-cloudius So are you suggesting to disable the load balancing if any of the known and alive nodes have it disabled?
> 
Or rather if all of them have it disabled.

--
			Gleb.

**Comment by tgrabiec** [comment]
> Or rather if all of them have it disabled.

But then in the mixed cluster the nodes which have HWLB disabled will appear as having 0 hitrate, and that will also cause imbalance of load.

**Comment by gleb-cloudius** [comment]
On Wed, Oct 23, 2019 at 11:31:14AM -0700, Tomasz Grabiec wrote:
> > Or rather if all of them have it disabled.
> 
> But then in the mixed cluster the nodes which have HWLB disabled will appear as having 0 hitrate, and that will also cause imbalance of load.
> 
> 
Why? Without this patch here a node always calculates hitrate, the flag
only controls if the info is actually used or not.

--
			Gleb.

**Comment by tgrabiec** [comment]
śr., 23 paź 2019, 21:00 użytkownik Gleb Natapov <notifications@github.com>
napisał:

> On Wed, Oct 23, 2019 at 11:31:14AM -0700, Tomasz Grabiec wrote:
> > > Or rather if all of them have it disabled.
> >
> > But then in the mixed cluster the nodes which have HWLB disabled will
> appear as having 0 hitrate, and that will also cause imbalance of load.
> >
> >
> Why? Without this patch here a node always calculates hitrate, the flag
> only controls if the info is actually used or not.
>

The flag also disables gossiping of CACHE_HITRATES, that's one of the
reasons we want it. So coordinators will see DISABLED. How do we treat that
when calculating hitrates? The old behavior was to interpret missing
CACHE_HITRATES as 0% hitrate.

>

**Comment by gleb-cloudius** [comment]
On Thu, Oct 24, 2019 at 10:10:09AM -0700, Tomasz Grabiec wrote:
> śr., 23 paź 2019, 21:00 użytkownik Gleb Natapov <notifications@github.com>
> napisał:
> 
> > On Wed, Oct 23, 2019 at 11:31:14AM -0700, Tomasz Grabiec wrote:
> > > > Or rather if all of them have it disabled.
> > >
> > > But then in the mixed cluster the nodes which have HWLB disabled will
> > appear as having 0 hitrate, and that will also cause imbalance of load.
> > >
> > >
> > Why? Without this patch here a node always calculates hitrate, the flag
> > only controls if the info is actually used or not.
> >
> 
> The flag also disables gossiping of CACHE_HITRATES, that's one of the
> reasons we want it. So coordinators will see DISABLED. How do we treat that
> when calculating hitrates? The old behavior was to interpret missing
> CACHE_HITRATES as 0% hitrate.
> 
What I am saying that what the patch does it incorrect. It should not
stop gossiping unless _all_ nodes have it disabled.

--
			Gleb.

**Comment by haaawk** [comment]
> On Thu, Oct 24, 2019 at 10:10:09AM -0700, Tomasz Grabiec wrote: śr., 23 paź 2019, 21:00 użytkownik Gleb Natapov ***@***.***> napisał: > On Wed, Oct 23, 2019 at 11:31:14AM -0700, Tomasz Grabiec wrote: > > > Or rather if all of them have it disabled. > > > > But then in the mixed cluster the nodes which have HWLB disabled will > appear as having 0 hitrate, and that will also cause imbalance of load. > > > > > Why? Without this patch here a node always calculates hitrate, the flag > only controls if the info is actually used or not. > The flag also disables gossiping of CACHE_HITRATES, that's one of the reasons we want it. So coordinators will see DISABLED. How do we treat that when calculating hitrates? The old behavior was to interpret missing CACHE_HITRATES as 0% hitrate.
> What I am saying that what the patch does it incorrect. It should not stop gossiping unless _all_ nodes have it disabled.
> […](#)
> -- Gleb.

How would the nodes know that everyone has HWLB disabled then @gleb-cloudius ? So far we were indicating it by stopping the gossip of hit rates.

**Comment by tgrabiec** [comment]
> What I am saying that what the patch does it incorrect. It should not stop gossiping unless _all_ nodes have it disabled.

Ok.

**Comment by tgrabiec** [comment]
> > On Thu, Oct 24, 2019 at 10:10:09AM -0700, Tomasz Grabiec wrote: śr., 23 paź 2019, 21:00 użytkownik Gleb Natapov _**@**_.***> napisał: > On Wed, Oct 23, 2019 at 11:31:14AM -0700, Tomasz Grabiec wrote: > > > Or rather if all of them have it disabled. > > > > But then in the mixed cluster the nodes which have HWLB disabled will > appear as having 0 hitrate, and that will also cause imbalance of load. > > > > > Why? Without this patch here a node always calculates hitrate, the flag > only controls if the info is actually used or not. > The flag also disables gossiping of CACHE_HITRATES, that's one of the reasons we want it. So coordinators will see DISABLED. How do we treat that when calculating hitrates? The old behavior was to interpret missing CACHE_HITRATES as 0% hitrate.
> > What I am saying that what the patch does it incorrect. It should not stop gossiping unless _all_ nodes have it disabled.
> > […](#)
> > -- Gleb.
> 
> How would the nodes know that everyone has HWLB disabled then @gleb-cloudius ? So far we were indicating it by stopping the gossip of hit rates.

You would need to introduce a new gossip state, HWLB_ENABLED, and check if all live endpoints have it enabled or not when making a decision at the coordinator side or whether to gossip CACHE_HITRATES. Perhaps we can have something like `gms::feature` which listens on gossip changes so that you don't have to actually compute this on each request.

**Comment by haaawk** [comment]
> Ah, we skipping the calculation if one of the nodes returned DISABLE, right. Not sure this is enough though. With a cluster of size 4 if one node has it disable but three others are not some requests will be balanced and some will not which may result in disproportional load on some nodes.

I have to admit that I don't understand your concern @gleb-cloudius. I don't think disabling HWLB on one node will cause other nodes to receive any more load. When HWLB is disabled on all nodes, the load will be balanced more or less equally among all of them. If HWLB is enabled then the load is distributed based on how hot are the caches. Disabling HWLB on a single node will cause to be more balanced as some requests will be load balanced equally to all nodes while others will still be distributed according to cache hotness. 

Since the goal is to disable HWLB on all nodes, we will get to a state when load is balanced more or less equally anyway. Disabling a single node moves the distribution of the load few steps closer to that final distribution. Why would that be a problem? If in equal distribution each node should receive X requests/s then turning HWLB on one node will increase load on nodes that currently receive less than X req/s and will decrease the load on nodes that currently receive more than X req/s.

@gleb-cloudius could you please explain a bit more what you meant by "disproportional load"? Disproportional to what? total load / number of nodes or hotness of the caches?

**Comment by gleb-cloudius** [comment]
On Fri, Oct 25, 2019 at 04:37:26AM -0700, Piotr Jastrzębski wrote:
> > Ah, we skipping the calculation if one of the nodes returned DISABLE, right. Not sure this is enough though. With a cluster of size 4 if one node has it disable but three others are not some requests will be balanced and some will not which may result in disproportional load on some nodes.
> 
> I have to admit that I don't understand your concern @gleb-cloudius. I don't think disabling HWLB on one node will cause other nodes to receive any more load. When HWLB is disabled on all nodes, the load will be balanced more or less equally among all of them. If HWLB is enabled then the load is distributed based on how hot are the caches. Disabling HWLB on a single node will cause to be more balanced as some requests will be load balanced equally to all nodes while others will still be distributed according to cache hotness. 
> 
> Since the goal is to disable HWLB on all nodes, we will get to a state when load is balanced more or less equally anyway. Disabling a single node moves the distribution of the load few steps closer to that final distribution. Why would that be a problem? If in equal distribution each node should receive X requests/s then turning HWLB on one node will increase load on nodes that currently receive less than X req/s and will decrease the load on nodes that currently receive more than X req/s.
> 
The problem is not that a node with disabled HWLB will stop distributing
a load according to cache hotness, this is indeed fine. The problem is
that it starts to advertise cache hit of zero, so _other_ nodes that
still do not have HWLB disabled will start to distribute load using
incorrect data. In affect they will stop sending load to a nodes with
disabled HWLB.

That said the data distributed over gossiper is used very briefly only
after a node reboots and not yet communicated with other nodes, but we
deemed closing this short window to be important, so I do not see the
reason to break it. 

--
			Gleb.

**Comment by haaawk** [comment]
> The problem is not that a node with disabled HWLB will stop distributing a load according to cache hotness, this is indeed fine. The problem is that it starts to advertise cache hit of zero, so _other_ nodes that still do not have HWLB disabled will start to distribute load using incorrect data. In affect they will stop sending load to a nodes with disabled HWLB. That said the data distributed over gossiper is used very briefly only after a node reboots and not yet communicated with other nodes, but we deemed closing this short window to be important, so I do not see the reason to break it.
> […](#)
> -- Gleb.

This patchset does not work the way you described.

If a node has disabled HWLB, it will not advertise cache hit of zero. It will advertise DISABLED.

_other_ nodes that still do not have HWLB disabled will see that DISABLED and whenever a node that advertises DISABLED is involved in a query, they won't use HWLB at all and balance load equally between all the replicas.

**Comment by haaawk** [comment]
I guess you're thinking about an older version of the code that does not understand DISABLED yet, right? In that case, you're right.

**Comment by haaawk** [comment]
> I guess you're thinking about nodes with an older version of the code that does not understand DISABLED yet, right? In that case, you're right.

The solution for this would be to first do a rolling upgrade with a new version that understands DISABLED but don't disable HWLB and only after that do rolling disable of HWLB but I'm not sure this is a practical solution.

**Comment by gleb-cloudius** [comment]
On Sun, Oct 27, 2019 at 01:26:02AM -0700, Piotr Jastrzębski wrote:
> > I guess you're thinking about nodes with an older version of the code that does not understand DISABLED yet, right? In that case, you're right.
> 
> The solution for this would be to first do a rolling upgrade with a new version that understands DISABLED but don't disable HWLB and only after that do rolling disable of HWLB but I'm not sure this is a practical solution.
> 
Looking at the patch again things looks even worse. table::get_hit_rate()
considers gossiper only if there were no previous communication with the
node. Otherwise it uses the data that was piggy backed to previous reply
(it is expensive to check gossiper on each request and data there is
less up-to-date). And since on a destination node HWLB calculation was
disabled the data will be incorrect (initialized to zero as far as I see).

--
			Gleb.

**Comment by haaawk** [comment]
It seems that the issue in the enterprise was fixed already so closing this PR.

**Pull request closed without merging**